cub/test/catch2_test_iterator.cu

```diff
@@ -0,0 +1,277 @@
+  CubDebugExit(cudaDeviceSynchronize());
+
+  InputIteratorT h_itrs[2]{d_in, d_in};
+  CHECK(cudaSuccess == cudaMemcpy(h_itrs, d_itrs, sizeof(InputIteratorT) * 2, cudaMemcpyDeviceToHost));
+
+  CHECK(0 == CompareDeviceResults(h_reference.data(), thrust::raw_pointer_cast(d_out.data()), h_reference.size()));
+  CHECK(d_in + 21 == h_itrs[0]);
+  CHECK(d_in == h_itrs[1]);
+
+  CHECK(cudaSuccess == cudaFree(d_itrs));
+#endif
+}
+
+CUB_TEST("Test constant iterator", "[iterator]", integral_types)
+{
+  using T                = c2h::get<0, TestType>;
+  const T base           = static_cast<T>(GENERATE(0, 99));
+  const auto h_reference = c2h::host_vector<T>{base, base, base, base, base, base, base, base};
+  test_iterator(cub::ConstantInputIterator<T>(base), h_reference);
+}
+
+CUB_TEST("Test counting iterator", "[iterator]", integral_types)
+{
+  using T                = c2h::get<0, TestType>;
+  const T base           = static_cast<T>(GENERATE(0, 99));
+  const auto h_reference = c2h::host_vector<T>{
+    static_cast<T>(base + 0),
+    static_cast<T>(base + 100),
+    static_cast<T>(base + 1000),
+    static_cast<T>(base + 10000),
+    static_cast<T>(base + 1),
+    static_cast<T>(base + 21),
+    static_cast<T>(base + 11),
+    static_cast<T>(base + 0)};
+  test_iterator(cub::CountingInputIterator<T>(base), h_reference);
+}
+
+using cache_modifiers =
+  c2h::enum_type_list<cub::CacheLoadModifier,
+                      cub::LOAD_DEFAULT,
+                      cub::LOAD_CA,
+                      cub::LOAD_CG,
+                      cub::LOAD_CS,
+                      cub::LOAD_CV,
+                      cub::LOAD_LDG,
+                      cub::LOAD_VOLATILE>;
+
+CUB_TEST("Test cache modified iterator", "[iterator]", types, cache_modifiers)
+{
+  using T                       = c2h::get<0, TestType>;
+  constexpr auto cache_modifier = c2h::get<1, TestType>::value;
+  constexpr int TEST_VALUES     = 11000;
+
+  c2h::host_vector<T> h_data(TEST_VALUES);
+  for (int i = 0; i < TEST_VALUES; ++i)
+  {
+    RandomBits(h_data[i]);
+  }
+  c2h::device_vector<T> d_data(h_data);
+
+  const auto h_reference = c2h::host_vector<T>{
+    h_data[0], h_data[100], h_data[1000], h_data[10000], h_data[1], h_data[21], h_data[11], h_data[0]};
+  test_iterator(
+    cub::CacheModifiedInputIterator<cache_modifier, T>(const_cast<const T*>(thrust::raw_pointer_cast(d_data.data()))),
+    h_reference);
+}
+
+template <typename T>
+struct TransformOp
+{
+  _CCCL_HOST_DEVICE T operator()(T input) const
+  {
+    T addend;
+    InitValue(INTEGER_SEED, addend, 1);
+    return input + addend;
+  }
+};
+
+CUB_TEST("Test transform iterator", "[iterator]", types)
+{
+  using T                   = c2h::get<0, TestType>;
+  constexpr int TEST_VALUES = 11000;
+
+  c2h::host_vector<T> h_data(TEST_VALUES);
+  for (int i = 0; i < TEST_VALUES; ++i)
+  {
+    InitValue(INTEGER_SEED, h_data[i], i);
+  }
+  c2h::device_vector<T> d_data(h_data.begin(), h_data.end());
```

Comment:
suggestion: the conventional pattern for generating random input in CUB tests is:
```suggestion
  c2h::device_vector<T> d_data(TEST_VALUES);
  c2h::gen(CUB_SEED(1), d_data);
  c2h::host_vector<T> h_data(d_data);
```
All the `RandomBits` and `InitValue` are legacy test facilities. Besides consistency, using `c2h::gen` uniformity across unit tests will help us later when we introduce test fuzzing. We'll use different seeds between invocations of the test to increase testing coverage.